### PR TITLE
[SPARK-7578] [ml] [doc] User guide for spark.ml Normalizer, IDF, StandardScaler

### PR DESCRIPTION
**Pull request by jkbradley**
Added user guide sections with code examples.
Also added small Java unit tests to test Java example in guide.

CC: @mengxr

### CODE REVIEW

**Comment by jkbradley** [comment]
@mengxr  Just rebased off of master, so this should be ready to review (with fewer PRs for the ml-features.md file now).

File: docs/ml-features.md [hunk 1]
```diff
@@ -18,30 +18,38 @@ This section covers algorithms for working with features, roughly divided into t
...
-val hashingTF = new HashingTF().setInputCol("words").setOutputCol("features").setNumFeatures(20)
-val featurized = hashingTF.transform(wordsDataFrame)
-featurized.select("features", "label").take(3).foreach(println)
+val wordsData = tokenizer.transform(sentenceData)
+val hashingTF = new HashingTF().setInputCol("words").setOutputCol("rawFeatures").setNumFeatures(20)
+val featurizedData = hashingTF.transform(wordsData)
+val idf = new IDF().setInputCol("rawFeatures").setOutputCol("features")
+val idfModel = idf.fit(featurizedData)
+val rescaledData = idfModel.transform(featurizedData)
+rescaledData.select("features", "label").take(3).foreach(println)
```
Comment (mengxr) [inline docs/ml-features.md]: minor: `take(3)` -> `collect()`

File: docs/ml-features.md [hunk 2]
```diff
@@ -618,5 +634,157 @@ indexedData = indexerModel.transform(data)
...
+{% highlight scala %}
+import org.apache.spark.ml.feature.Normalizer
+import org.apache.spark.mllib.util.MLUtils
+
+val data = MLUtils.loadLibSVMFile(sc, "data/mllib/sample_libsvm_data.txt")
+val dataFrame = sqlContext.createDataFrame(data)
+val normalizer = new Normalizer().setInputCol("features").setOutputCol("normFeatures")
+
+// Normalize each Vector using $L^2$ norm.
+val l2NormData = normalizer.transform(dataFrame, normalizer.p -> 2)
```
Comment (mengxr) [inline docs/ml-features.md]: `p = 2` is the default value. Maybe we can use `setP(1.0)` for normalizer, and then `transform(..., normalizer.p -> inf)` in to demo how to use extra params.

File: docs/ml-features.md [hunk 3]
```diff
@@ -618,5 +634,157 @@ indexedData = indexerModel.transform(data)
...
+{% highlight python %}
+from pyspark.mllib.util import MLUtils
+from pyspark.ml.feature import Normalizer
+
+data = MLUtils.loadLibSVMFile(sc, "data/mllib/sample_libsvm_data.txt")
+dataFrame = sqlContext.createDataFrame(data)
+normalizer = Normalizer(inputCol="features", outputCol="normFeatures")
+
+# Normalize each Vector using $L^2$ norm.
+l2NormData = normalizer.transform(dataFrame, {normalizer.p:2.0})
```
Comment (mengxr) [inline docs/ml-features.md]: space after `:`

File: docs/ml-features.md [hunk 4]
```diff
@@ -618,5 +634,157 @@ indexedData = indexerModel.transform(data)
...
+<div data-lang="scala">
+{% highlight scala %}
+import org.apache.spark.ml.feature.StandardScaler
+import org.apache.spark.mllib.util.MLUtils
+
+val data = MLUtils.loadLibSVMFile(sc, "data/mllib/sample_libsvm_data.txt")
+val dataFrame = sqlContext.createDataFrame(data)
+val scaler = new StandardScaler()
+  .setInputCol("features")
+  .setOutputCol("normFeatures")
```
Comment (mengxr) [inline docs/ml-features.md]: minor: `normFeatures` -> `scaledFeatures` or `stdFeatures`?

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaHashingTFSuite.java [hunk 5]
```diff
@@ -63,17 +63,22 @@ public void hashingTF() {
...
-      .setOutputCol("features")
+      .setOutputCol("rawFeatures")
       .setNumFeatures(numFeatures);
-    DataFrame featurized = hashingTF.transform(wordsDataFrame);
-    for (Row r : featurized.select("features", "words", "label").take(3)) {
+    DataFrame featurizedData = hashingTF.transform(wordsData);
+    IDF idf = new IDF().setInputCol("rawFeatures").setOutputCol("features");
+    IDFModel idfModel = idf.fit(featurizedData);
+    DataFrame rescaledData = idfModel.transform(featurizedData);
+    for (Row r : rescaledData.select("features", "label").take(3)) {
```
Comment (mengxr) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaHashingTFSuite.java]: `take(3)` -> `collect()`?

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaNormalizerSuite.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  @After
+  public void tearDown() {
+    jsc.stop();
+    jsc = null;
+  }
+
+  @Test
+  public void regexTokenizer() {
```
Comment (mengxr) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaNormalizerSuite.java]: `normalizer`

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaNormalizerSuite.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      new VectorIndexerSuite.FeatureData(Vectors.dense(1.0, 4.0))
+    );
+    DataFrame dataFrame = jsql.createDataFrame(jsc.parallelize(points, 2),
+      VectorIndexerSuite.FeatureData.class);
+    Normalizer normalizer = new Normalizer()
+      .setInputCol("features")
+      .setOutputCol("normFeatures");
+
+    // Normalize each Vector using $L^2$ norm.
+    DataFrame l2NormData = normalizer.transform(dataFrame, normalizer.p().w(2));
```
Comment (mengxr) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaNormalizerSuite.java]: Call `count()` to trigger the job.

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaNormalizerSuite.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    // Normalize each Vector using $L^\infty$ norm.
+    DataFrame lInfNormData =
+      normalizer.transform(dataFrame, normalizer.p().w(Double.POSITIVE_INFINITY));
```
Comment (mengxr) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaNormalizerSuite.java]: ditto

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaStandardScalerSuite.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  @After
+  public void tearDown() {
+    jsc.stop();
+    jsc = null;
+  }
+
+  @Test
+  public void regexTokenizer() {
```
Comment (mengxr) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaStandardScalerSuite.java]: `standardScaler`

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaStandardScalerSuite.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      .setInputCol("features")
+      .setOutputCol("normFeatures")
+      .setWithStd(true)
+      .setWithMean(false);
+
+    // Compute summary statistics by fitting the StandardScaler
+    StandardScalerModel scalerModel = scaler.fit(dataFrame);
+
+    // Normalize each feature to have unit standard deviation.
+    DataFrame scaledData = scalerModel.transform(dataFrame);
```
Comment (mengxr) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaStandardScalerSuite.java]: `count()`

File: docs/ml-features.md (hunk 1 above)
Comment (jkbradley) [inline docs/ml-features.md]: I thought about that, but was worried users would copy this into their own code and try to print out a huge dataset.  What do you think?

**Comment by jkbradley** [comment]
@mengxr  Updated.  Thanks!

**Comment by mengxr** [comment]
LGTM. Merged into master and branch-1.4. Thanks!

**Pull request closed without merging**